src/python/zquantum/qaoa/recursive_qaoa.py

```diff
@@ -43,156 +48,127 @@ def __init__(
+        # and will be reset to these original values at the end of the last recursion.
+        self._qubit_map = _create_default_qubit_map(n_qubits)
+        self._original_cost_hamiltonian = cost_hamiltonian
+        self._nit = 0
+        self._nfev = 0
+        self._histories: Dict[str, List[HistoryEntry]] = defaultdict(list)
+        self._histories["history"] = []
+
+    def _minimize(
         self,
-        cost_hamiltonian: IsingOperator,
-        qubit_map: Dict[int, List[int]] = None,
-    ) -> List[Tuple[int, ...]]:
+        cost_function_factory: Callable[[IsingOperator, Ansatz], CostFunction],
+        initial_params: np.ndarray,
+        keep_history: bool = False,
+    ) -> OptimizeResult:
         """Args:
-            cost_hamiltonian: Hamiltonian representing the cost function.
-            qubit_map: A dictionary that maps qubits in reduced Hamiltonian back to original qubits, used
-                for subsequent recursions. (Not for the first recursion.)
-                Example:
-                    `qubit_map = {0: [2, -1], 1: [3, 1]]}
-                        Keys are the original qubit indice.
-                        1st term of inner list is qubit the index of tuple to be mapped onto,
-                        2nd term is if it will be mapped onto the same value or opposite of the qubit it
-                            is being mapped onto.
-                        In the above qubit_map, the original qubit 0 is now represented by the opposite
-                            value of qubit 2, and the original qubit 1 is now represented by the value of
-                            qubit 3.
+            cost_function_factory: function that generates CostFunction objects given the provided ansatz
+                and cost_hamiltonian.
+            initial_params: initial parameters used for optimization
+            keep_history: flag indicating whether history of cost function
+                evaluations should be recorded.
 
         Returns:
-            The solution(s) to recursive QAOA as a list of tuples; each tuple is a tuple of bits.
+            OptimizeResult with the added entry of:
+                opt_solutions (List[Tuple[int, ...]]): The solution(s) to recursive QAOA as a list of tuples;
+                    each tuple is a tuple of bits.
         """
 
-        n_qubits = count_qubits(change_operator_type(cost_hamiltonian, QubitOperator))
-
-        if self._n_c >= n_qubits or self._n_c <= 0:
-            raise ValueError(
-                "n_c needs to be a value less than number of qubits and greater than 0."
-            )
-
-        if qubit_map is None:
-            qubit_map = _create_default_qubit_map(n_qubits)
-
         ansatz = copy(self._ansatz)
 
         if hasattr(ansatz, "_cost_hamiltonian"):
-            ansatz._cost_hamiltonian = cost_hamiltonian
+            ansatz._cost_hamiltonian = self._cost_hamiltonian
```

Comment:
I think this whole logic should be removed.
1. I think the responsibility of providing the same cost Hamiltonian to both ansatz and cost function should lie on the user. Also, I've seen a case, where people deliberately were providing different Hamiltonians, so there is such a usecase as well.
2. Removing it would also allow us to delete the whole custom logic for X-ansatz.
3. There's no need for using `ansatz._cost_hamiltonian` since `ansatz.cost_hamiltonian` should also work just fine.

Reply:
That's an excellent point!
So I think this makes 1st point invalid :) 
Regarding 2nd point, I would remove the logic for the X ansatz anyway.
And point 3 is still valid :)